saleor/product/views.py

```diff
@@ -115,5 +118,18 @@ def category_index(request, path, category_id):
     if actual_path != path:
         return redirect('product:category', permanent=True, path=actual_path,
                         category_id=category_id)
-    return TemplateResponse(request, 'category/index.html',
-                            {'category': category})
+    products = (products_with_details(user=request.user)
+                .filter(categories__name=category)
+                .order_by(DEFAULT_SORT))
+    product_filter = ProductFilter(
+        request.GET, queryset=products, category=category)
+    products_paginated = get_paginator_items(
+        product_filter.qs, PAGINATE_BY, request.GET.get('page'))
+    products_and_availability = list(products_with_availability(
+        products_paginated, request.discounts, request.currency))
+    ctx = {'category': category, 'filter': product_filter,
+           'products': products_and_availability,
+           'products_paginated': products_paginated,
+           'sort_by_choices': SORT_BY_FIELDS,
+           'show_pagination': len(products) > PAGINATE_BY}
```

Comment:
`len(products)` makes additional queries and evaluates the whole queryset of products. This makes the pagination pointless.

Reply:
Done